text/0058-unseal-spancontext.md

```diff
@@ -0,0 +1,117 @@
+
+> SpanContexts are immutable.
+
+only applies to the part of the SpanContext that is publicly exposed on the API level.
+Actual vendor implementations could have additional mutable properties or
+actually have the first access to the `TraceId` property generate the TraceId lazily. Vendors must take care not to break any assumptions by doing that. For example, if a language allows copying a SpanContext, vendors should better put all mutable data behind an immutable reference so that all copies remain in a consistent state.
+
+## Internal details
+
+The changes in the initial summary of this OTEP shall be applied to the specification at
+https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#spancontext.
+
+Language libraries that used a "final" or "sealed" keyword or something similar in their API definition of SpanContext must remove it.
+
+The reference SDK implementation will continue to use the API's SpanContext implementation without deriving from it.
+Additionally, all public APIs of the SpanContext must be overridable
+(they must not be public data fields in languages where these are not overridable,
+languages that require an explicit marker like `virtual` to make an API overridable must add it, etc.).
+
+Existing propagators
+(which should be implemented in the SDK package or a separate package)
+can continue to directly construct a SpanContext.
+API implementations that want to use a custom SpanContext implementation must either
+handle having a plain SpanContext as parent
+and having additional properties stripped upon extracting
+or they provide their own propagator implementations too.
+
+## Trade-offs and mitigations
+
+The decision to have SpanContext as a non-abstract class in the API results from the goal to keep compatibility.
+The cleaner alternative would be to have SpanContext be an interface like all other API classes.
+
+Since the implementation of propagators is on the SDK level, all public APIs could be removed from the SpanContext
+so that it becomes an object that is completely opaque (to everyone but the propagator implementations).
+This would enable even more flexibility or at least spare vendors from adding boilerplate code that, e.g., calculates TraceIds
+if there is no such concept in their own system.
+
+In a (currently hypothetical) C++ implementation of OpenTelemetry (and possibly some other languages),
+having the SpanContext be a "polymorphic", potentially derived class is a big difference to
+the old SpanContext that might be implemented as a simple data structure with a few data fields.
+These languages could opt to make SpanContext effectively a "smart pointer".
+That is, the SpanContext could be implemented as a final class containing nothing but a reference to the actual SpanContextData (which is not final) and non-virtual functions that delegate to the corresponding virtual functions of the SpancontextData class.
```

Comment:
This is a huge performance hit for languages like C++, because you require refcounting for the SpanContext now. I think this is a big mistake, and some companies may refuse to use this. We need to propose a solution that does not force heap allocation + refcounting for C++ (maybe golang as well where you need this to be an interface now).

Reply:
That huge performance hit is already incurred for Spans though.

Reply:
Yes, which is a problem, but you are doubling that.

Reply:
I think I misread your first comment a bit. You were complaining about the reference counting! This is not necessarily required, if you don't need the smart pointer to be copyable, you can use a unique_ptr-like zero-cost abstraction. Also, I wrote "could opt to". But thinking about it now, it seems in my head we already had a C++-implemenation were we need to care about minimizing breaking changes. In a new implementation you would probably just return a unique_ptr<SpanContext> and have users use `->` instead of `.` I guess I should just remove that paragraph.

If Span implements SpanContext, that would be a bit of a problem because then who is responsible for deleting the span is different between SpanContexts extracted from remote requests and SpanContexts that refer to inproc Spans. That could be solved with a custom smart pointer (nearly zero-cost, I'm thinking about a `bool shouldDelete` member) though.

Anyway, I feel this conversation is drifting too much into C++-specifics. I guess I should really remove that paragraph.